internal/dlv/delve.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package dlv
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/go-delve/delve/cmd/dlv/cmds"
+)
+
+// MainWithArgs is a type representing a function that takes command-line arguments as input and returns an exit code.
+type MainWithArgs func(args []string) int
+
+const (
+	// envNoDebug is an environment variable used to redirect the control flow to a new delve instance
+	// or to the program. It is set as environment variable the first time the flow pass in the function
+	envNoDebug = "JUJU_NO_DLV_DEBUG"
+)
+
+// NewDlvRunner wraps a MainWithArgs function to enable debugging.
+//
+// It works in two phases:
+// At first call, it launches the delve command `exec` with the current binary. However, before the
+// call it set a environment variable DELVE_ANYTHING_NO_DEBUG.
+// At the second call, DELVE_ANYTHING_NO_DEBUG is set, so it just return the "normal" main.
+func NewDlvRunner(opts ...Option) func(main MainWithArgs) MainWithArgs {
+
+	// Apply configuration option.
+	config := Config{}
+	config.apply(opts...)
+
+	logger := config.logger()
+
+	return func(main MainWithArgs) MainWithArgs {
+		if _, exists := os.LookupEnv(envNoDebug); exists {
+			return main
+		}
+		if err := os.Setenv(envNoDebug, "1"); err != nil {
+			logger.Printf("Failed to set env %q: %v", envNoDebug, err)
+			logger.Printf("Starting without debug mode...")
+			return main
+		}
+
+		// Run with delve
+		return func(args []string) int {
+
+			// Prepare dlv commands with expected options and command
+			command := args[0]
+			dlvArgs := append(config.args(), "exec", command, "--")
+			dlvArgs = append(dlvArgs, args[1:]...)
+			dlvCmd := cmds.New(false)
+			dlvCmd.SetArgs(dlvArgs)
+
+			// Start "sidecars" (go routine required to make some commands works... Like fixing permissions on listening
+			// socket
+			config.runSidecars()
```

Comment:
I agree on the principle (and I though about it when writing the code). 

However, doing more here feels like over engineering:

* We are not talking about millions of goroutine. goroutine leak causes issues when they are a lot.
* It is a functionnality to deal with edge cases. That is a "allows hack functionnality", so being too rigid on it may discard some valuable future usecase
* This code is designed to work only in debug usecase. Not in prod. I think it grant us the possibility to be a bit lax on some edge case.

Open to discuss it. I have already an idea of how make it cleaner, I just think it didn't need to, because, since it's for debug usecase (non prod), I follow readability/flexibility over cleanliness and purity.